### PR TITLE
SMAPIv3 `zfs-vol` driver (aka zfs-ng gen2)

### PR DESCRIPTION
**Pull request by ydirson**
This is a 2nd gen prototype based on #11, incrementally rebuilt piece by piece

### CODE REVIEW

File: plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    def create(self, dbg, sr_uuid, configuration, name, description):
+        log.debug('{}: SR.create: config={}, sr_uuid={}'.format(
+            dbg, configuration, sr_uuid))
+
+        # 2 ways to create a SR:
+        # - from an existing zpool (manually created for complex configs)
+        # - from a "devices" config string (comma-separated list of devices)
+
+        if 'zpool' in configuration:
+            if 'devices' not in configuration:
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py]: it is if 'devices' in configuration I think

**Comment by gthvn1** [comment]
This first part looks good :+1:

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+from xapi.storage.common import call
+
+MOUNT_ROOT = '/var/run/sr-mount'
+
+def pool_mountpoint(dbg, pool_name):
+    cmd = "zfs get mountpoint -H -o value".split() + [ pool_name ]
+    return call(dbg, cmd).strip()
+
+def pool_create(dbg, pool_name, devs):
+    cmd = [
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: nit and it is ok for me but why not splitting cmd like others?

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    cmd = "zpool get -Hp -o value size".split() + [ sr_path ]
+    return int(call(dbg, cmd))
+
+def pool_get_free_space(dbg, sr_path):
+    # size is returned in bytes
+    cmd = "zpool get -Hp -o value free".split() + [ sr_path ]
+    return int(call(dbg, cmd))
+
+def vdi_create(dbg, zvol_path, size_mib):
+    cmd = [ 'zfs', 'create',
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: same here

File: plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py (hunk 1 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py]: yes, that was corrected in a later fixup - everything has been squashed now

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py (hunk 2 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: I pondered that, but I felt it as more clear to keep the options together with their arguments, which only left "zpool create" to split, so that would look quite weird.  But then I left the `.split()` for `destroy` so yes it is in fact lacking consistency

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py [hunk 4]
```diff
@@ -22,12 +22,23 @@ def zvol_find_snap_path(dbg, pool_name, snap_id):
...
 
-def _zvol_ensure_destroyable(dbg, vol_name):
+def zvol_ensure_destroyable(dbg, vol_name):
     cmd = "zfs list -Hp -o name,origin".split()
-    pattern = vol_name + "@"
+
+    # a zvol depends on a snapshot if its origin is == to that
+    # snapshot, and depends on another zvol if its origin is a
+    # snapshot of that zvol
+    if "@" in vol_name:
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: "@" in the name means that it is a snapshot right? So you are speaking about snapshot that depends of another snapshots here?

Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: no, this code checks whether other zvols depend on `vol_name`, and the check is different whether  `vol_name` is a snapshot (will appear directly as origin) or a zvol (where we look for snapshots based on it)

File: plugins/volume/org.xen.xapi.storage.zfs-ng/plugin.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        return {
+            "plugin": "zfs-ng",
+            "name": "ZFS Volume plugin",
+            "description": ("This plugin manages ZFS volumes"),
+            "vendor": "None",
+            "copyright": "(C) 2021-2024 Vates",
+            "version": "3.0",
+            "required_api_version": "5.0",
+            "features": [
+                "VDI_CREATE",
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/plugin.py]: shouldn't we add *SR_* features as well?

File: plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        # 2 ways to create a SR:
+        # - from an existing zpool (manually created for complex configs)
+        # - from a "devices" config string (comma-separated list of devices)
+
+        if 'zpool' in configuration:
+            if 'devices' in configuration:
+                log.error('"zpool" specified, "devices" should not be used')
+                raise Exception('"zpool" specified, "devices" should not be used')
+
+            # FIXME do we reject pools not under MOUNT_ROOT?
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py]: With #18 we choose where we are mounting the pool so we don't need to reject pools not under MOUNT_ROOT because we will force the mount point. I think to pool created with legacy mode by a user. We can deal with that.

File: plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+            # FIXME validate existence of pool first?
+            pool_name = configuration['zpool']
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py]: Maybe it can be a good idea. But if we don't we will have an issue later when trying to attach the SR. So is it better to check now that we can import and mount the pool instead of do it later in the SR.attach. Maybe yes. We will need #18 for that :)

File: plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+            # then becomes the sole way to designate the SR
+            configuration["orig-devices"] = configuration['devices']
+            del configuration['devices']
+            configuration["zpool"] = pool_name
+
+        else:
+            log.error('devices config must have "zpool" or "devices"')
+            raise Exception('devices config must have "zpool" or "devices"')
+
+        # FIXME this assumes zpool is mounted/attached
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py]: Again with #18 we can try import/mount. And in fact it answers to the previous FIXME where we finally need to check that the zpool provided by the user exist and can be mounted.

File: plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        return configuration
+
+    def destroy(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        zfsutils.pool_destroy(dbg, meta["zpool"])
+
+    def attach(self, dbg, configuration):
+        log.debug('{}: SR.attach: config={}'.format(dbg, configuration))
+
+        # ZFS automagically attaches a pool to a mountpoint on
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py]: fixed with #18

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        proc = subprocess.Popen(
+            cmd_args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            close_fds=True)
+        stdout, stderr = proc.communicate()
+        if error and proc.returncode != expRc:
+            log.error('%s: %s exitted with code %s: %s',
+                      dbg, " ".join(cmd_args), proc.returncode, stderr)
+            if ntries > 1 and "is busy" in stderr:
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: if returncode is 2 it means that the command line is invalid so we probably don't need to retry for this one.

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+###
+
+def vol_get_used(dbg, vol_name):
+    # size is returned in bytes
+    cmd = "zfs get -Hp -o value used".split() + [ vol_name ]
+    return int(call(dbg, cmd))
+
+def vol_create(dbg, zvol_path, size_mib):
+    cmd = ("zfs create".split() + [zvol_path]
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: for my info why parenthesis are needed here?

**Review by gthvn1 — COMMENTED** [review]
Still need to review that part about destroying snapshots. WIP...

File: plugins/volume/org.xen.xapi.storage.zfs-ng/plugin.py (hunk 5 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/plugin.py]: I deliberately did not add those whose for which I could not see any impact omitting them.
I must say I have trouble reading the ocaml code checking for features, which from an orbital bird's view looks like it should check them.

File: plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py (hunk 6 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py]: I'd leave the FIXME's to be removed by #18, so things will be clear later when reading history

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py (hunk 11 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: python needs to know the expression continues on next line

File: plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+@util.decorate_all_routines(util.log_exceptions_in_function)
+class Implementation(DefaultImplementation):
+    "Volume driver to provide volumes from zvol's"
+
+    def create(self, dbg, sr, name, description, size, sharable):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        pool_name = meta["zpool"]
+
+        with VolumeContext(self.callbacks, sr, 'w') as opq:
+            # FIXME how should we choose image format?
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py]: RAW seems appropriate here. Do you think about adding an new format like BDEV (Block Device) ?

File: plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+        cb = self.callbacks
+        need_destroy_clone = False
+        with VolumeContext(cb, sr, 'w') as opq:
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    vdi = db.get_vdi_by_id(key)
+                    is_snapshot = vdi.volume.snap
+                    if is_snapshot:
+                        snap_name = zfsutils.zvol_find_snap_path(dbg, pool_name, vdi.volume.id)
+                        zfsutils.zpool_log_state(dbg, "before destroy {}".format(snap_name), pool_name)
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py]: nit: I think it is "before destroying"

File: plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py (hunk 7 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py]: That somewhat converges with my comment on #18 about having a helper to know a zpool is already imported: that one would need its own helper to determine if a zpool exists even if not imported - or we can just define this mode of operation to mean "an imported zpool".

There is also the question whether we protect ourselves against getting passed a zpool what would already be in use, which OTOH expecting "an exported zpool" would answer.

**Comment by gthvn1** [comment]
I tested the creation of a VDI, snaphost it, clone it and delete the inital VDI. It seems ok. while testing I followed the code and it looks good to me :+1:

File: plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py (hunk 13 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py]: in my mind it is "before `destroy` operation" ;)

File: plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py (hunk 12 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py]: Yes RAW seems good, finally, as we've ruled out the nested-cow use-case anyway.

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py (hunk 4 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: code disappeared since

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+            # then becomes the sole way to designate the SR
+            configuration["orig-vdev"] = configuration['vdev']
+            del configuration['vdev']
+            configuration["zpool"] = pool_name
+
+        else:
+            log.error('devices config must have "zpool", "vdev", or "device"')
+            raise Exception('devices config must have "zpool", "vdev", or "device"')
+
+        # FIXME this assumes zpool is mounted/attached
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:77]: maybe we can call attach (with version #18) ?

File: plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+                    volume.vsize = zfsutils.vol_get_size(dbg, vol_name)
+                    if volume.vsize != size:
+                        log.debug("%s: VDI.create adjusted requested size %s to %s",
+                                  dbg, size, volume.vsize)
+                    db.update_volume_vsize(volume.id, volume.vsize)
+
+            vdi_uri = self.callbacks.getVolumeUriPrefix(opq) + vdi_uuid
+
+        return {
+            'key': vdi_uuid,    # FIXME check this
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py]: the only requirement is that it should be unique? At least the common volume.py is doing the same. But I don't know if it is ok.

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        pool_name = meta["zpool"]
+        cb = importlib.import_module('zfs-vol').Callbacks()
+        with VolumeContext(cb, sr, 'r') as opq:
+            with cb.db_context(opq) as db:
+                vdis = db.get_all_vdis()
+                all_custom_keys = db.get_all_vdi_custom_keys()
+
+            for vdi in vdis:
+                # TODO: handle this better
+                # _vdi_sanitize(vdi, opq, db, cb)
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py]: Shouldn't we call it instead of commenting? If we need to handle this better maybe it should be done in another patch that will modify `_vdi_sanitize`?

File: plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+            'uuid': vdi_uuid,
+            'name': name,
+            'description': description,
+            'read_write': True,
+            'virtual_size': size,
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py]: shouldn't be `volume.vsize` that seems to be the adjusted size?

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py (hunk 16 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py]: It's not that easy as this function is not applicable, but something similar is part of last weeks work still in fixup commits, see ab2dc908966e6724ab8179eeed5271113314d245 and later

File: plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py (hunk 17 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py]: right, good catch!

Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py]: fixed

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py (hunk 14 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:77]: I'm not a big fan of the idea - we could maybe introduce a couple of helpers to be called by both, though.
That continues to advocate for #18 to become "squash" commits at some point

File: plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py (hunk 10 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py]: fixed

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py (hunk 14 above)
Comment (gthvn1) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:77]: Sure for #18, you can take it to be part of your fixes (if this is what you mean). I will close it as implemented elsewhere. And I agree that calling attach is not great and probably adding an helper like "ensure_mounted" that tries to get the mountpoint and if not tries to import it once is better. Or also modified `import` to check if it is already mounted and let `import` be idempotent. In this case we can call it from here to ensure that it is mounted and raise an exception if it is not mounted and cannot be mounted.

**Comment by ydirson** [comment]
Rerolled:
- applied fixups
- sync'd commit messages for new `zls-vol` name
- split #18 commit into `SR.attach` and `SR.detach` ones

**Comment by gthvn1** [comment]
LGTM

**Comment by ydirson** [comment]
Pushed:
- removal of FIXME tags for `key` and `psize`
- fix of `psize` on `Volume.create` (we're creating thin volumes now)
- rewording of plugin description

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+        # - from an existing zpool (manually created for complex configs)
+        # - from a "device" config string (comma-separated list of devices)
+        # - from a "vdev" config string (vdev specification suitable for `zfs create`)
+
+        if 'zpool' in configuration:
+            if 'device' in configuration or 'vdev' in configuration:
+                log.error('"zpool" specified, "device" or "vdev" should not be used')
+                raise Exception('"zpool" specified, "device" or "vdev" should not be used')
+
+            # FIXME validate existence of pool first?
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:38]: I can propose a patch for this one based on zfs import maybe. So we can merge your patch and fix this later.

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+            configuration["orig-vdev"] = configuration['vdev']
+            del configuration['vdev']
+            configuration["zpool"] = pool_name
+
+        else:
+            log.error('devices config must have "zpool", "vdev", or "device"')
+            raise Exception('devices config must have "zpool", "vdev", or "device"')
+
+        # FIXME this assumes zpool is mounted/attached
+        mountpoint = zfsutils.pool_mountpoint(dbg, pool_name)
```
Comment (gthvn1) [inline plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:78]: maybe at least try expect and print an easy to understand error "expected zpool to be mounted" before raising the error? I'm not sure if raise directly is easy to read in log.

**Review by gthvn1 — APPROVED** [review]
minor comments. LGTM !

File: plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py (hunk 15 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py]: yes, does not look useful to dig deeper at this stage

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py (hunk 16 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py]: My bad, that one being in `sr.py` not `volume.py` I missed it.  Addressed in upcoming push.

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py (hunk 19 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:78]: For the prototype it should be enough to get the limitation listed in the doc.  We have a work item for this post-proto.

File: plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py (hunk 18 above)
Comment (ydirson) [reply plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:38]: Post-proto work item

**Comment by ydirson** [comment]
New update with:
- avoid single-use temp psize/fsize variables
- remove remaining useless FIXME's, replaced by better info
- fixed some scoping errors with statements outside of the `with` blocks they belong too (python is too permissive) (but no runtime impact today)
- renamed and moved `zfsvol_vdi_sanitize` to `zfsutils` to add the missing call in `sr.ls`
- don't fail a `sr-scan` when we cannot locate zvol for a snapshot, just skip it in the output, and log an error

**Comment by gthvn1** [comment]
As the fixme about the zpool already mounted will be done later it seems reasonable for me to merge your PR.
Great job :+1:

**Review by gthvn1 — APPROVED** [review]
